gensim/models/poincare.py

```diff
@@ -0,0 +1,637 @@
+        self.distance_gradients_v = distance_gradients_v
+
+        self.distance_gradients_computed = True
+
+    def compute_loss(self):
+        """Compute and store loss value for the given batch of examples."""
+        if self.loss_computed:
+            return
+        self.compute_distances()
+
+        self.loss = -np.log(self.exp_negative_distances[0] / self.Z).sum()  # scalar
+        self.loss_computed = True
+
+
+class PoincareKeyedVectors(KeyedVectors):
+    """
+    Class to contain vectors and vocab for the PoincareModel training class,
+    can be used to perform operations on the vectors such as vector lookup, distance etc.
+
+    """
+    @staticmethod
+    def poincare_dist(vector_1, vector_2):
+        """Return poincare distance between two vectors."""
+        norm_1 = np.linalg.norm(vector_1)
+        norm_2 = np.linalg.norm(vector_2)
+        euclidean_dist = np.linalg.norm(vector_1 - vector_2)
+        if euclidean_dist == 0.0:
+            return 0.0
+        else:
+            return np.arccosh(
+                1 + 2 * (
+                    (euclidean_dist ** 2) / ((1 - norm_1 ** 2) * (1 - norm_2 ** 2))
+                )
+            )
+    # TODO: Add other KeyedVector supported methods - most_similar, etc.
+
+
+class PoincareModel(utils.SaveLoad):
+    """
+    Class for training, using and evaluating Poincare Embeddings described in https://arxiv.org/pdf/1705.08039.pdf
+
+    The model can be stored/loaded via its `save()` and `load()` methods, or stored/loaded in the word2vec
+    format via `wv.save_word2vec_format()` and `KeyedVectors.load_word2vec_format()`.
+
+    Note that training cannot be resumed from a model loaded via `load_word2vec_format`, if you wish to train further,
+    use `save()` and `load()` methods instead.
+
+    """
+    def __init__(
+        self, train_file, size=50, alpha=0.1, negative=10, iter=50,
+        workers=1, epsilon=1e-5, burn_in=10, burn_in_alpha=0.01, encoding='utf8', seed=0):
+        """
+        Initialize and train a Poincare embedding model from a file of transitive closure relations.
+
+        Parameters
+        ----------
+        train_file : str
+            Path to tsv file containing relation pairs.
+        size : int, optional
+            Number of dimensions of the trained model, defaults to 50.
+        alpha : float, optional
+            learning rate for training, defaults to 0.1.
+        negative : int, optional
+            Number of negative samples to use, defaults to 10.
+        iter : int, optional
+            Number of iterations (epochs) over the corpus, defaults to 50.
+        workers : int, optional
+            Number of threads to use for training the model, defaults to 1.
+        epsilon : float, optional
+            Constant used for clipping embeddings below a norm of one, defaults to 1e-5.
+        burn_in : int, optional
+            Number of epochs to use for burn-in initialization (0 means no burn-in), defaults to 0.
```

Comment:
defaults to 10 according to method signature

Reply:
let's not write default values at all, because this is useless duplication.

Reply:
Removed